README.md
# STP CLIENT

Version: 2.7
Last update: 2015-12-14

### Java Dependencies:

The app was compiled with Java 1.7.0_76.
If you plan to insert deals into a database, ensure the JRE and database have the same architecture (32/64 bits).

Main jars: 
- STPClient.jar,
- com.ibm.mq.jar,
- com.ibm.mqjms.jar,
- jbossall-client.jar,

Enable emails:
- activation.jar,
- mail.jar,

Logging utility: 
- log4j.jar,

ODBC Driver, as required:
- sqlite-jdbc-3.7.2.jar,

QuickFix/J:
- mina-core-1.1.7.jar
- mina-filter-ssl-1.1.7.jar
- proxool-0.9.1.jar
- quickfixj-all-1.5.3.jar
- sleepycat-je_2.1.30.jar
- slf4j-api-1.6.3.jar
- slf4j-log4j12-1.6.3.jar


### Queue Connection and XSLT

This application connects to the IBM MQ queue and automatically downloads the messages. The details of the connection need
to be set in the properties file. The same property file will be used to configure all options described below.

The message downloaded is a special XML file called FinXML; once downloaded, the FinXML file is transformed into 
another format using XSLT (http://www.w3schools.com/xsl/).
XSLT requires a stylesheet file; the stylesheet file informs the XSLTransformer how the new file should look like. 
For example, tidyXml.xsl will create a nicer and simpler xml file, while csv.xsl will create a comma separated line.

These are the properties. You need to set the Stylesheet, while the other properties will be provided by Integral.

	STP.STYLESHEET=config/tidyXml.xsl
	STP.IP=xxx.xxx.xxx.xxx
	STP.PORT=yyyy
	STP.QUEUE_MANAGER=
	STP.QUEUE_NAME=XXX.PROD.QUEUE
	STP.CHANNEL=XXXX.PROD.CHANNEL
	STP.USERNAME=
	STP.PASSWORD=
	STP.CERT_NAME=
	STP.SSL_CIPHER_SUITE=


### Database Connection

Optionally, and only if the XSLT creates an xml file with just 1 level of depth - like tidyXml.xsl does -, the message
can be saved to a database via the standard ODBC/JDBC connection properties.

The stylesheet must be written with the database schema in mind: the application will use the output 
xml file from the XSL Transformation as input for the SQL query, for example, say the XSLT file outputs:

	<myMessage>
		<dealid>FXI123456789</dealid>
		<tradedate>2014-10-07</tradedate>
		<rate>1.65897</rate>
	</myMessage>

Then the sql query will be:

	INSERT INTO myMessage (dealid, tradedate, rate) VALUES ('FXI123456789', '2014-10-07', 1.65897);


Relevant properties are below. These are the standard ODBC connection details.

	DB.SWITCH=ON
	DB.JDBC_DRIVER=org.sqlite.JDBC
	DB.URL=jdbc:sqlite:STPMessages.db
	DB.USERNAME=
	DB.PASSWORD=


### Email

Optionally, an email can be sent for every new message. If the stylesheet file is creating a XML file, then
you can reformat the body of the email using another stylesheet file, and you can also customize the subject using
as keywords the nodes in the XML message.
You can add an address book with key value pairs to send the email to specific receivers depending on one or more specific factors. 
For example, your address book can look like this:

	1000BRKZ/1000BRKZle=team1@whatnot.com
	1000BRKZ/1000BRKZLE2=team2@whatnot.com

and the email will be sent to the specific team if the key specified in *ADDRESS_BOOK_KEY* matches any key in the address book, 
otherwise it will be sent to the default which is what is set in *TO*.

You can also add a logo image. Below the properties.


	EMAIL.SWITCH=ON
	EMAIL.FROM=stp@integral.com
	EMAIL.TO=fabio@integral.com
	EMAIL.ADDRESS_BOOK=config/AddressBook.txt
	EMAIL.ADDRESS_BOOK_KEY=<OrgID>,/,<OrgLE>
	EMAIL.HOST=out.int.com
	EMAIL.PORT=25
	EMAIL.AUTH=FALSE
	EMAIL.USERNAME=
	EMAIL.PASSWORD=
	EMAIL.STYLESHEET=config/tidyEmail.xsl
	EMAIL.SUBJECT=<TradeID>, ,<OrgID>, ,<BuySell>, ,<BaseCcy>,/,<TermCcy>, ,<BaseAmt>, ,<ValueDate>, vs ,<CptyID>, @ ,<AllInRate>
	EMAIL.IMAGE=config/integral.jpg

### White Label Email

In addition, you can send an email to the counterparty. You will require a file with key-value pairs for the email 
addresses. An example is below; the key is the Counterparty ID:

	8000BRKZ=treasury@bigbank.com
    1000BRKZ=joe@abcbank.com,sue@abcbank.com

You can specify a custom stylesheet, image and subject so that the deal is from the client's perspective. 
Set property *WL.SIDE_INDICATOR* to help customize the subject of the email with the direction from the perspective
of the WL client. The value to this property should be the field name that indicates the side followed by the indicators 
used to identify a buy or a sell (eg: Buy/Sell; B/S; 1/2; etc.). The 3 values should be comma separated (eg:Side,Buy,Sell).
Once this property is configured, adding eg: \<Side\> to the WL.SUBJECT will return the direction from the perspective of the
WL client. You must set in WL.CPTY_ID the node in the XML file that identifies the WL client organization.
You are required to set EMAIL.SWITCH to 'ON'.

	WL.SWITCH=ON
    WL.CPTY_ID=CptyID
    WL.STYLESHEET=config/wlEmail.xsl
    WL.SUBJECT=<TradeID>, ,<CptyUser>, ,<Side>,s, ,<BaseCcy>,/,<TermCcy>, ,<BaseAmt>, ,<ValueDate>, @ ,<AllInRate>
    WL.SIDE_INDICATOR=Side,Buy,Sell
    WL.CLIENTS_EMAILS=config/WLClients.txt
    WL.IMAGE=config/wlIntegral.jpg


### Save individual deals to a file

Optionally, you can save each deal to an individual file. You can use specific keywords (Date, TradeID
and Counter) to customize the file name.
The counter reset every day and starts at 0 if not specified. To enable the counter, you must specify the file location.

	FILE.SWITCH=ON
	FILE.PATH=C:/Users/ghirardellof/IdeaProjects/STPclient
	FILE.DATE_FORMAT=yyyyMMdd_HHmmss
	FILE.COUNTER=config/counter
	FILE.NAME=<TradeID>,.,<Date>,.xml

### FIX messaging

You can convert the FinXML message into a FIX message. The stylesheet must output a file with parameter 'tag'.
Example output file:

	<myMessage>
		<TradeId tag="17">FXI123456789</TradeId>
		<TradeDate tag="64">2014-10-07</TradeDate>
		<AllInRate tag="44">1.65897</AllInRate>
	</myMessage>

Then the FIX message TradeCaptureReport (AE) will output something like this:

	[...]|17=FXI123456789|64=2014-10-07|44=1.65897|[...]
	
The FIX configuration file is a separate file whose location needs to be specified in the properties. 

	QFJ.SWITCH=OFF
	QFJ.CONFIG_FILE=config/qfj.properties
	

### Running the app

Below is the command to run the application. This assumes you have all the required libraries in lib/,
and the properties file, the xsl files and a log4j.xml file in config/.

	java -cp lib/* -Dlog4j.configuration=file:config/log4j.xml STPClient config/app.properties


### Logging

Messages can be saved into a separate log file. This is accomplished with the logging utility 'log4j' by setting the proper
appender to the category "MyTrade".
Log4j requires its own configuration file. The category that returns the application logging is 'STPClient'.




# CHANGE LOG

### 2.7 (2015-12-14)

- Added Address Book functionality to the Email feature.


### 2.6 (2015-06-30)

- Added Thread.sleep() to lower CPU load.

### 2.5 (2015-06-03)

- Incorporated STPConnector.jar into STPClient.jar.

### 2.4 (2015-05-18)

- Added new property *WL.IMAGE* to insert a custom image on WL emails.
- Added new property *WL.SIDE_INDICATOR* to help customize the subject of the email with the direction from the perspective
of the WL client. The value to this property should be the field name that indicates the side followed by the indicators 
used to identify a buy or a sell (eg: Buy/Sell; B/S; 1/2; etc.). The 3 values should be comma separated (eg:Side,Buy,Sell).
Once this property is configured, adding eg: <Side> to the WL.SUBJECT will return the direction from the perspective of the 
WL client.
- The stylesheet for sending FIX messages uses now parameters to identify the tag, rather than the "field_#" convention (eg: <TradeID_17>).
For example, to send *\<TradeId\>* as tag #17, use *\<TradeId tag="17"\>*.
- Logging configuration file log4j.xml should be updated with the new class names, as follows:
    - *com.integral* --> *IBMMQClient* (to grab connection logging and the raw finXML file);
    - *FabSTPClient* --> *STPClient* (to grab all application logging).
- STPClient.jar now features the version in the filename.


### 2.3 (2015-05-14)

- Recompiled *STPConnector.jar* to return a Document object instead of String, thus the stylesheet should return a XML file.
- Renamed *FabSTPClient* to *STPClient*. This is important to launch the app.
- Stylesheet files meant to return plain text should be enclosed in a *\<root\>* node; this is so that the app recognizes 
these Document objects are actually plain text files. An example is *csv.xsl*.
- Renamed properties *EMAIL.TITLE, WL.TITLE, FILE.TITLE* to *EMAIL.SUBJECT, WL.SUBJECT, FILE.NAME*.
- Added class *WriteFile* to separate the function to save the message to an individual file from class STPClient.


CHANGELOG.md
### 2.7 (2015-12-14)

- Added Address Book functionality to the Email feature.


### 2.6 (2015-06-30)

- Added Thread.sleep() to lower CPU load.

### 2.5 (2015-06-03)

- Incorporated STPConnector.jar into STPClient.jar.

### 2.4 (2015-05-18)

- Added new property *WL.IMAGE* to insert a custom image on WL emails.
- Added new property *WL.SIDE_INDICATOR* to help customize the subject of the email with the direction from the perspective
of the WL client. The value to this property should be the field name that indicates the side followed by the indicators 
used to identify a buy or a sell (eg: Buy/Sell; B/S; 1/2; etc.). The 3 values should be comma separated (eg:Side,Buy,Sell).
Once this property is configured, adding eg: <Side> to the WL.SUBJECT will return the direction from the perspective of the 
WL client.
- The stylesheet for sending FIX messages uses now parameters to identify the tag, rather than the "field_#" convention (eg: <TradeID_17>).
For example, to send *\<TradeId\>* as tag #17, use *\<TradeId tag="17"\>*.
- Logging configuration file log4j.xml should be updated with the new class names, as follows:
    - *com.integral* --> *IBMMQClient* (to grab connection logging and the raw finXML file);
    - *FabSTPClient* --> *STPClient* (to grab all application logging).
- STPClient.jar now features the version in the filename.


### 2.3 (2015-05-14)

- Recompiled *STPConnector.jar* to return a Document object instead of String, thus the stylesheet should return a XML file.
- Renamed *FabSTPClient* to *STPClient*. This is important to launch the app.
- Stylesheet files meant to return plain text should be enclosed in a *\<root\>* node; this is so that the app recognizes 
these Document objects are actually plain text files. An example is *csv.xsl*.
- Renamed properties *EMAIL.TITLE, WL.TITLE, FILE.TITLE* to *EMAIL.SUBJECT, WL.SUBJECT, FILE.NAME*.
- Added class *WriteFile* to separate the function to save the message to an individual file from class STPClient.


src/STPClient.java
import org.apache.log4j.Logger;
import org.w3c.dom.Document;

import javax.xml.transform.*;
import javax.xml.transform.dom.DOMResult;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Properties;

public class STPClient {
    private static Logger log = Logger.getLogger("STPClient");
    private static Logger logTrade = Logger.getLogger("MyTrade");
    private static final String VERSION = "2.7";
    private static final String COMPILE_DATE = "2015-12-14";
    private static final String JDK_VERSION = "1.7.0_76";

    private static DB db;
    private static Email email;
    private static WL wl;
    private static QFJ qfj;
    private static WriteFile writeFile;


    // STP connection variables
    private static String STP_STYLESHEET;
    private static String STP_IP;
    private static String STP_PORT;
    private static String STP_QUEUE_MANAGER;
    private static String STP_QUEUE_NAME;
    private static String STP_CHANNEL;
    private static String STP_USERNAME;
    private static String STP_PASSWORD;
    private static String STP_SSL_CIPHER_SUITE;

    // DB Properties variables
    private static String DB_SWITCH;
    private static String DB_JDBC_DRIVER;
    private static String DB_URL;
    private static String DB_USERNAME;
    private static String DB_PASSWORD;

    // Email Properties variables
    private static String EMAIL_SWITCH;
    private static String EMAIL_FROM;
    private static String EMAIL_TO;
    private static String EMAIL_ADDRESS_BOOK;
    private static String EMAIL_ADDRESS_BOOK_KEY;
    private static String EMAIL_HOST;
    private static String EMAIL_PORT;
    private static String EMAIL_AUTH;
    private static String EMAIL_USERNAME;
    private static String EMAIL_PASSWORD;
    private static String EMAIL_STYLESHEET;
    private static String EMAIL_SUBJECT;
    private static String EMAIL_IMAGE;

    // White Label Properties variables
    private static String WL_SWITCH;
    private static String WL_CPTY_ID;
    private static String WL_STYLESHEET;
    private static String WL_SUBJECT;
    private static String WL_SIDE_INDICATOR;
    private static String WL_CLIENTS_EMAILS;
    private static String WL_IMAGE;

    // Save To File variables
    private static String FILE_SWITCH;
    private static String FILE_PATH;
    private static String FILE_DATE_FORMAT;
    private static String FILE_COUNTER;
    private static String FILE_TITLE;

    // QuickFix/J variables
    private static String QFJ_SWITCH;
    private static String QFJ_CONFIG_FILE;


    public STPClient(String properties) {
        readProperties(properties);

        // check if there's a db connection setup
        if (DB_SWITCH.equalsIgnoreCase("ON")) {
            db = new DB(DB_JDBC_DRIVER, DB_URL, DB_USERNAME, DB_PASSWORD);
        }

        // check if there's a email setup
        if (EMAIL_SWITCH.equalsIgnoreCase("ON")) {
            email = new Email(EMAIL_FROM, EMAIL_TO, EMAIL_ADDRESS_BOOK, EMAIL_ADDRESS_BOOK_KEY, EMAIL_HOST, EMAIL_PORT,
                    EMAIL_USERNAME, EMAIL_PASSWORD, EMAIL_AUTH, EMAIL_STYLESHEET, EMAIL_SUBJECT, EMAIL_IMAGE);
        }

        //check if there's a WL email setup
        if (WL_SWITCH.equalsIgnoreCase("ON") && EMAIL_SWITCH.equalsIgnoreCase("ON") ) {
            wl = new WL(email, WL_CPTY_ID, WL_STYLESHEET, WL_SUBJECT, WL_SIDE_INDICATOR, WL_CLIENTS_EMAILS, WL_IMAGE);
        }

        // check if deals need to be saved to an individual file
        if (FILE_SWITCH.equalsIgnoreCase("ON")) {
            writeFile = new WriteFile(FILE_DATE_FORMAT, FILE_COUNTER, FILE_TITLE, FILE_PATH);
        }

        // check if deals need to be sent via FIX
        if (QFJ_SWITCH.equalsIgnoreCase("ON")) {
            qfj = new QFJ(QFJ_CONFIG_FILE);
        }

        // launch the STP listener
        IBMMQClient ibmmqClient = new IBMMQClient(STP_IP, STP_PORT, STP_QUEUE_MANAGER, STP_QUEUE_NAME,
                STP_CHANNEL, STP_USERNAME, STP_PASSWORD, STP_SSL_CIPHER_SUITE,
                new IBMMQClient.OnMessageReceived() {
                    @Override
                    //here 'messageReceived' method-event is implemented
                    public void messageReceived(String tradeID, String finXML) {
                        publish(tradeID, finXML);
                    }
                });

        ibmmqClient.login();
    }


    public static void main(String args[]) {
        log.info("STPClient version " + VERSION + ". Compiled with JDK version " + JDK_VERSION + " on " + COMPILE_DATE + ".");
        // print out the config file
        try {
            String line = new String(Files.readAllBytes(Paths.get(args[0])));
            log.info("Reading configuration file " + args[0] + ":\n" + line);
        } catch (IOException e) {
            log.error("[ERR001] ", e);
        }

        STPClient stpClient = new STPClient(args[0]);

        while (true) {
            try {
                Thread.sleep(Long.MAX_VALUE);
            }
            catch (InterruptedException ie){
                log.info("STPClient has been interrupted.");
            }
        }

    }

    private void readProperties(String configFile ){
        InputStream input = null;
        try {
            Properties prop = new Properties();
            input = new FileInputStream(configFile);

            // load a properties file
            prop.load(input);

            // get STP Connection properties
            STP_STYLESHEET = prop.getProperty("STP.STYLESHEET");
            STP_IP = prop.getProperty("STP.IP");
            STP_PORT = prop.getProperty("STP.PORT");
            STP_QUEUE_MANAGER = prop.getProperty("STP.QUEUE_MANAGER");
            STP_QUEUE_NAME = prop.getProperty("STP.QUEUE_NAME");
            STP_CHANNEL = prop.getProperty("STP.CHANNEL");
            STP_USERNAME = prop.getProperty("STP.USERNAME");
            STP_PASSWORD = prop.getProperty("STP.PASSWORD");
            //STP_CERT_NAME = prop.getProperty("STP.CERT_NAME");
            STP_SSL_CIPHER_SUITE = prop.getProperty("STP.SSL_CIPHER_SUITE");

            // get the DB properties
            DB_SWITCH = prop.getProperty("DB.SWITCH", "OFF");
            DB_JDBC_DRIVER = prop.getProperty("DB.JDBC_DRIVER");
            DB_URL = prop.getProperty("DB.URL");
            DB_USERNAME = prop.getProperty("DB.USERNAME");
            DB_PASSWORD = prop.getProperty("DB.PASSWORD");

            // get the Email Properties
            EMAIL_SWITCH = prop.getProperty("EMAIL.SWITCH", "OFF");
            EMAIL_FROM = prop.getProperty("EMAIL.FROM");
            EMAIL_TO = prop.getProperty("EMAIL.TO", "");
            EMAIL_ADDRESS_BOOK = prop.getProperty("EMAIL.ADDRESS_BOOK", "");
            EMAIL_ADDRESS_BOOK_KEY = prop.getProperty("EMAIL.ADDRESS_BOOK_KEY", "");
            EMAIL_HOST = prop.getProperty("EMAIL.HOST");
            EMAIL_PORT = prop.getProperty("EMAIL.PORT");
            EMAIL_AUTH = prop.getProperty("EMAIL.AUTH");
            EMAIL_USERNAME = prop.getProperty("EMAIL.USERNAME");
            EMAIL_PASSWORD = prop.getProperty("EMAIL.PASSWORD");
            EMAIL_STYLESHEET = prop.getProperty("EMAIL.STYLESHEET", "");
            EMAIL_SUBJECT = prop.getProperty("EMAIL.SUBJECT", "");
            EMAIL_IMAGE = prop.getProperty("EMAIL.IMAGE", "");

            // get the White Label properties
            WL_SWITCH = prop.getProperty("WL.SWITCH","OFF");
            WL_CPTY_ID = prop.getProperty("WL.CPTY_ID");
            WL_STYLESHEET = prop.getProperty("WL.STYLESHEET","");
            WL_SUBJECT = prop.getProperty("WL.SUBJECT","");
            WL_SIDE_INDICATOR = prop.getProperty("WL.SIDE_INDICATOR"," , , ");
            WL_CLIENTS_EMAILS = prop.getProperty("WL.CLIENTS_EMAILS");
            WL_IMAGE = prop.getProperty("WL.IMAGE", "");

            // get Save to File properties
            FILE_SWITCH = prop.getProperty("FILE.SWITCH", "OFF");
            FILE_PATH = prop.getProperty("FILE.PATH");
            FILE_DATE_FORMAT = prop.getProperty("FILE.DATE_FORMAT", "");
            FILE_COUNTER = prop.getProperty("FILE.COUNTER", "");
            FILE_TITLE = prop.getProperty("FILE.NAME");

            // get QuickFix/J  properties
            QFJ_SWITCH = prop.getProperty("QFJ.SWITCH", "OFF");
            QFJ_CONFIG_FILE = prop.getProperty("QFJ.CONFIG_FILE");

        }
        catch (IOException e) {
            log.error("[ERR002] - Error reading properties file. ", e);
        }
        finally {
            if (input != null) {
                try {
                    input.close();
                }
                catch (IOException e) {
                    //do nothing
                }
            }
        }
    }

    /*
     * This method is called once a new STP message has arrived
     */
    public void publish(String tradeID, String finXML) {
        Document doc = XslTransform(finXML);

        logTrade.info(docToString(doc));
        if (db != null) {
            db.insert(tradeID, doc);
        }
        if (email != null) {
            email.send(tradeID, doc);
        }
        if (wl != null) {
            wl.sendEmailToWLClient(tradeID, doc);
        }
        if (writeFile != null) {
            writeFile.writeToFile(tradeID, docToString(doc));
        }
        if (qfj != null) {
            qfj.sendTradeCaptureReport(doc);
        }
    }


    public Document XslTransform(String finXML) {
        StreamSource xml = new StreamSource(new StringReader(finXML));
        StreamSource xsl = new StreamSource(new File(this.STP_STYLESHEET).getAbsoluteFile());
        Document doc = null;
        try {
            TransformerFactory tFactory = TransformerFactory.newInstance();
            Transformer transformer = tFactory.newTransformer(xsl);
            DOMResult domResult = new DOMResult();
            transformer.transform(xml, domResult);
            doc = (Document)domResult.getNode();
        }
        catch (TransformerFactoryConfigurationError | TransformerException e) {
            e.printStackTrace();
        }
        return doc ;
    }

    public static String docToString(Document doc) {
        try {
            StringWriter sw = new StringWriter();
            TransformerFactory tf = TransformerFactory.newInstance();
            Transformer transformer = tf.newTransformer();
            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");

            if (doc.getFirstChild().getNodeName().equalsIgnoreCase("root")) {
                transformer.setOutputProperty(OutputKeys.METHOD, "text");
            }
            //transformer.setOutputProperty(OutputKeys.INDENT, "yes");
            //transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");

            transformer.transform(new DOMSource(doc), new StreamResult(sw));
            return sw.toString();
        } catch (Exception ex) {
            throw new RuntimeException("Error converting to String", ex);
        }
    }
}


